Make COSE deserialization more robust

robin-nitrokey opened the pull request:
<s>Previously, COSE keys could only be deserialized if the map entries were in the correct order.</s>  With this patch, the deserialization is made more robust (and also more concise) by introducing a helper struct, RawPublicKey, that is used as an intermediate step during the deserialization.  It also adds support for parsing keys without an algorithm field.

daringer commented:
lgtm, tests ran w/o issues

nickray commented on src/cose.rs:
```diff
@@ -133,13 +155,98 @@ impl From<TotpPublicKey> for PublicKey {
 
-#[derive(Clone, Debug, Eq, PartialEq)]
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+#[serde(into = "RawPublicKey")]
```
Why is there no `with = "RawPublicKey"` here?

robin-nitrokey replied:
There is no `with` argument for containers.  Do you mean `from`/`try_from`?  I chose to manually implement `Deserialize` to make error handling easier, but we can also add our own error type and use `TryFrom`.

nickray commented:
@robin-nitrokey can you add more context to this? Did you run into a specific issue?

The reason we're currently not doing this is related to [CTAP2 canonical CBOR encoding form][http://fidoalliance.org/specs/fido-v2.1-rd-20210309/fido-client-to-authenticator-protocol-v2.1-rd-20210309.html#ctap2-canonical-cbor-encoding-form]. Now I agree that, while we need to be sure we emit "canonical" CBOR, it might be helpful to be less strict when receiving. Is there a kind of "non-exhaustive" issue?

nickray commented on src/cose.rs:
```diff
@@ -163,7 +283,20 @@ impl PublicKeyConstants for EcdhEsHkdf256PublicKey {
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+#[serde(into = "RawPublicKey")]
```
Same comment as above, wouldn't we want `Deserialize` via `with = "RawPublicKey"` too?

robin-nitrokey commented:
@nickray A user reported an issue with `fido2luks` that is fixed by this patch, see https://github.com/Nitrokey/nitrokey-3-firmware/issues/177.  Besides the order of map keys, the old implementation also requires the algorithm field that should be optional.

robin-nitrokey commented:
I wasn’t aware that the standard also discourages accepting standard but non-canonical CBOR.  If you want to keep this behavior, the approach in this PR could still be used to simplify the serde trait implementations.

sosthene-nitrokey commented:
When reviewing this PR i checked the COSE spec and it says that canonical encoding rules applies do not apply to public keys.

https://datatracker.ietf.org/doc/html/rfc8152#section-14

>   -  The restriction applies to the encoding of the Sig_structure, the
      Enc_structure, and the MAC_structure.

But the FIDO2 spec appears to be more constrained than that.

nickray commented:
Yes, you need to distinguish upstream COSE canonical encoding and "CTAP2 canonical CBOR encoding form" unfortunately. I think it's better to stick with the strict (applicable) standard and consider issues arising from this a bug - unless you are strongly of a different opinion.

robin-nitrokey commented:
I’ve rebased this PR and rewrote it based on the previous discussion.  This means we still require canonical encoding, but also keep the simplified de-/serialization with `RawPublicKey` and the support for optional algorithms.

robin-nitrokey commented:
I found one final problem during the implementation:  The CTAP spec requires us to ignore unknown map keys.  The documentation for  [`MapAccess`](https://docs.rs/serde/latest/serde/de/trait.MapAccess.html) does not require `next_key` to be followed by `next_value`, so I assumed that just calling `next_key` until we see a key that we know should work.

Yet it does not work – `cbor_smol` [requires][cbor-1] a `next_value` call after each `next_key`.  But we cannot call `next_value` because we do not know which type this unknown value has, and `cbor_smol` [requires][cbor-2] the type to be known at compile time.

[cbor-1]: https://github.com/nickray/cbor-smol/blob/899b74f352582bd15dac1587e7163c99506ccb45/src/de.rs#L221-L238
[cbor-2]: https://github.com/nickray/cbor-smol/blob/899b74f352582bd15dac1587e7163c99506ccb45/src/de.rs#L287

So I think we have to stick with the current implementation here, failing for unknown keys until we have found all keys we want.

sosthene-nitrokey commented:
So we'd need to implement support for `deserialize_ignored_any` in `cbor-smol` ?

sosthene-nitrokey commented:
There is already a pull request for it: https://github.com/nickray/cbor-smol/pull/5

robin-nitrokey commented:
Yes, with that we should be able to avoid the problem.

sosthene-nitrokey commented:
A PR that works: https://github.com/nickray/cbor-smol/pull/6

robin-nitrokey commented:
Nice!  Here is an improved implementation using `deserialize_any` based on this PR:  https://github.com/trussed-dev/ctap-types/pull/22
